src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs

```diff
@@ -250,29 +250,43 @@ public void UpdateTrackedDocument(DocumentUri uri, SourceText newSourceText)
         // Find the matching document from the LSP solutions.
         foreach (var (workspace, lspSolution, isForked) in lspSolutions)
         {
-            var document = await lspSolution.GetTextDocumentAsync(textDocumentIdentifier, cancellationToken).ConfigureAwait(false);
-            if (document != null)
+            var documents = await lspSolution.GetTextDocumentsAsync(textDocumentIdentifier.DocumentUri, cancellationToken).ConfigureAwait(false);
+            if (documents.Length > 0)
             {
-                // Record metadata on how we got this document.
-                var workspaceKind = document.Project.Solution.WorkspaceKind;
-                _requestTelemetryLogger.UpdateFindDocumentTelemetryData(success: true, workspaceKind);
-                _requestTelemetryLogger.UpdateUsedForkedSolutionCounter(isForked);
-                _logger.LogDebug($"{document.FilePath} found in workspace {workspaceKind}");
+                // We have at least one document, so find the one in the right project context
+                var document = documents.FindDocumentInProjectContext(textDocumentIdentifier, (sln, id) => sln.GetRequiredTextDocument(id));
 
-                // If we found the document in a non-misc workspace, also attempt to remove it from the misc workspace
-                // if it happens to be in there as well.
-                if (_lspMiscellaneousFilesWorkspaceProvider is not null && !await _lspMiscellaneousFilesWorkspaceProvider.IsMiscellaneousFilesDocumentAsync(document, cancellationToken).ConfigureAwait(false))
+                if (_lspMiscellaneousFilesWorkspaceProvider is not null)
                 {
-                    try
+                    // If we started with multiple documents and didn't have specific context information, it's possible we picked a miscellaneous files document when
+                    // we could have picked a real one.
+                    if (documents.Length > 1 && await _lspMiscellaneousFilesWorkspaceProvider.IsMiscellaneousFilesDocumentAsync(document, cancellationToken).ConfigureAwait(false))
                     {
-                        await _lspMiscellaneousFilesWorkspaceProvider.TryRemoveMiscellaneousDocumentAsync(uri).ConfigureAwait(false);
+                        // Pick a different one; our choice here is arbitrary, since if we had a specified context in the first place we would have picked the right one.
+                        document = documents.First(d => d != document);
```

Comment:
Just scrutinizing the assumption in calling `First` here. We do expect the items in `documents` to be distinct, so this won't throw for `documents.Length > 1`.

Reply:
Correct; if they're not distinct something is quite wrong. (And it'll be "obvious" what the problem was if this throws.)